### PR TITLE
Closes #26995: add sorting of wallpapers on wallpapers onboarding tool

### LINKED ISSUE #26995: wallpapers onboarding filtering
Follow up for [issues/26750](https://github.com/mozilla-mobile/fenix/issues/26750)

Currently only classic wallpapers are shown. We want to have 3 seasonal first, if available, the rest – classic.

[design](https://www.figma.com/file/d7l1iGabQYqnLcnQbHDGzM/Wallpaper-2.0-%2B-Onboarding-v106?node-id=7946%3A223536)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21573)

### PR DESCRIPTION
**Pull request by mavduevskiy**
Filtering of wallpapers for wallpaper onboarding tool.

We are given a list of wallpapers, that consists of classic and seasonal collections. They come from the server, so theoretically the numbers inside each group could be any. Wrote an extension function that has a goal to provide a sorted list, which under ideal conditions would have 3 seasonal wallpapers first and 2 classic after. 

Edge cases I could come up with are: when it's all classic, or all seasonal, or nothing at all, or 1 classic or 1 seasonal, or 5 seasonal and 1 classic, or 1 seasonal and 5 classic.

image

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #26995

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/onboarding/WallpaperOnboardingDialogFragment.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/WallpaperOnboardingDialogFragment.kt:133]:
```suggestion
        // The desired amount of seasonal wallpapers inside of the wallpaper selector.
```

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:36]: Rather than looping the wallpaper several times, I am wondering what if we first called `groupByDisplayableCollection` which will provide us a mapping of the collection and the list of wallpapers associated to their respective collection. You can then determine if the count of each group matches the desired number of seasonal wallpaper or not, and then take the appropriate number of wallpapers from each group to get the final list of wallpapers to display.

Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:36]: That was actually my first approach as well, but I didn't like the implementation. Sadly, I have removed it. 

What I didn't like about it: we can't access the classic wallpapers from the map directly, because the access is done through an object we didn't create. I had to look for it and store it in a separate parameter in a for loop. 

And there might be multiple seasonal groups, so we would have to iterate through them as well. It looked like having collections created more problems than brought benefits. In the context of this particular selector there are just seasonal and classic wallpapers, and we can differentiate them in a list as well.

And lastly, it looks like we can't escape having two loops, because we have two groups of not guaranteed length. One loop might satisfy a single condition (let's say, 3 seasonal) but in case we have just 1 classic for some reason, we would have go back to add another seasonal, that we haven't added yet.

I am open to improving the sorting though. This just looked like the optimal approach given the possibly any input from the backend.

Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:36]: I don't really know how the wallpapers are structure. So, I am just gonna think out loud here and see if there is any possible approaches here that will make this code simpler. 
1. Is possible to go through the list of wallpaper and construct 2 lists one for seasonal and another for classic wallpapers?
2. After you have the 2 lists, you can check the count of each list and from there take the appropriate number of wallpapers from each.

Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:37]: okay, got it! that actually a much easier approach. updated

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:42]:
```suggestion
    val result = mutableListOf(Wallpaper.Default)
    val classicWallpapers = mutableListOf<Wallpaper>()
    val seasonalWallpapers = mutableListOf<Wallpaper>()
    
    for (wallpaper in this) {
```
Generally, we want a new line before and after a `for` statement block to separate out and make it easier to read. Otherwise, all our code will just be squished together with no clear logical separation from a reader's perspective.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:54]:
```suggestion
        result.addAll(classicWallpapers.take((maxSize - 1) - seasonalWallpapers.size))
```
My initial reaction was to wonder where `maxSize` was coming from. I think this code would be clearer if we didn't alias `THUMBNAILS_SELECTION_COUNT` as `maxSize`, but I imagine you did this because the name will take up a lot of space, but I would argue that we can just move `.take()` onto the next line.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:37]: Thanks, this does make much more easier to understand and maintain.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:36]:
```suggestion
/**
 * Returns a list of wallpapers to display in the wallpaper onboarding.
 *
 * The ideal scenario is to return a list of wallpaper in the following order: 1 default, 3 seasonal and 
 * 2 classic wallpapers, but in case where there are less than 3 seasonal wallpapers, the remaining
 * wallpapers are filled by classic wallpapers. If we have less than 6 wallpapers, return all the available 
 * seasonal and classic wallpapers.
 */
```
We do want a line that describes what the function does.  Otherwise, I helped reword your comment to make it flow more naturally from an english speaking perspective.

File: app/src/main/java/org/mozilla/fenix/onboarding/WallpaperOnboardingDialogFragment.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/onboarding/WallpaperOnboardingDialogFragment.kt:131]:
```suggestion
        // The number of wallpaper thumbnails to display.
        const val THUMBNAILS_SELECTION_COUNT = 6
```

Let's add a comment for this while we are at it.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:37]:
```suggestion
fun List<Wallpaper>.getWallpapersForOnboarding(): List<Wallpaper> {
```
We aren't grouping anything like a `Map` so I think we shouldn't use the same naming as above.

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt:58]:
```suggestion
    fun `GIVEN two collections of appropriate size WHEN fetched for onboarding THEN result contains 3 seasonal and 2 classic`() {
```
s/enough/appropriate/g
s/not enough/insufficient/g

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt:72]:
```suggestion
    fun `GIVEN five collections of insufficient size WHEN fetched for onboarding THEN result contains 3 seasonal and 2 classic`() {
```

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (gabrielluong) [inline app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt:68]: Instead of using `assert`, we generally do `assertTrue` or `assertFalse` to be more explicit.

**Review by gabrielluong — COMMENTED** [review]
I am gonna defer to @MatthewTighe for the final review, but I think it looks good.

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (MatthewTighe) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt:68]: IIRC `assert` won't actually trigger test failures in some cases either, as it's a keyword rather than part of the JUnit framework

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:66]: This could also potentially be simplified by doing something like the following:

```
val defaultAndSeasonal = listOf(Wallpaper.Default) + seasonalWallpapers.take(SEASONAL_WALLPAPERS_COUNT)

return defaultAndSeasonal + classicWallpapers.take(THUMBNAILS_SELECTION_COUNT - defaultAndSeasonal.size)
```

**Comment by mavduevskiy** [comment]
@gabrielluong 
addressed your comments, thanks!
no worries, I think I sometimes forget to squash before commit during the recent merging rush. I like to keep separate commits when I work on a branch to see the progress, if I squash now, I would erase some of the history about the work progress. I will do that before commiting, you have my word :)

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (MatthewTighe) [inline app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:31]: This looks like just a formatting change. I think in general we want to avoid doing this, but I am fine keeping it here for simplicity and so we don't hit it again in the future

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:66]: that fails the `GIVEN no items in classic collection and some seasonal WHEN grouped for onboarding THEN result contains all seasonal` test! We might face a case where we have to take more seasonal, then just 3, when there is not enough classic. it's an edge case, but given the api, I would like to write something that considers all the possible cases.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (mavduevskiy) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:31]: ops, that was not intentional. but yes, I agree, makes sense to keep it

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (gabrielluong) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: @mavduevskiy Just a bump on this. This is mostly an English wording nit.

File: app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/settings/wallpaper/Extensions.kt:66]: Ah, good catch. Let's just keep it as is then since it does a good job addressing all these edge cases

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (mavduevskiy) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: I have updated to the suggested version, is there anything else I should do?

Comment (gabrielluong) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: `/g` meant global here so we'll want to do the same for the other tests.

Comment (mavduevskiy) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: oh, got it! what does `/s` stand for?

Comment (gabrielluong) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: Sorry, `s/enough/appropriate/g` was a sed notation which is a linux command for search and transforming strings. You can read this as s/<old>/<new> where <old> will be replaced by <new> `/g` will be global and similar to what you see with regular expressions.

https://linuxhint.com/50_sed_command_examples/#s1

Comment (mavduevskiy) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: thanks! I didn't understand your comment properly, I did get only the suggested part.
updated!

**Review by MatthewTighe — APPROVED** [review]
Tested locally, LGTM!

**Comment by MatthewTighe** [comment]
Let's land this before https://github.com/mozilla-mobile/fenix/pull/27018 to make rebasing earlier. Looks like the commits still need to be squashed but otherwise I think this is good to go

**Comment by mavduevskiy** [comment]
@gabrielluong @MatthewTighe 
FYI squashed before landing!

File: app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt
Comment (gabrielluong) [reply app/src/test/java/org/mozilla/fenix/settings/wallpaper/ExtensionsTest.kt]: @mavduevskiy Sorry looks like the `<old>` and `<new>` were hidden in the markdown. I updated the comment in case it will be helpful.

**Pull request merged**
